modules/github-bots/sdk/bot.go

```diff
@@ -0,0 +1,83 @@
+package sdk
+
+import (
+	"context"
+	"encoding/json"
+	"log/slog"
+	"runtime/debug"
+
+	"github.com/chainguard-dev/clog"
+	"github.com/chainguard-dev/clog/gcp"
+	"github.com/chainguard-dev/terraform-infra-common/modules/github-events/schemas"
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/google/go-github/v61/github"
+	"github.com/kelseyhightower/envconfig"
+)
+
+type Bot interface {
+	Name() string
+}
+
+func Serve(b Bot) {
+	var env struct {
+		Port int `envconfig:"PORT" default:"8080" required:"true"`
+	}
+	if err := envconfig.Process("", &env); err != nil {
+		clog.Fatalf("failed to process env var: %s", err)
+	}
+
+	slog.SetDefault(slog.New(gcp.NewHandler(slog.LevelInfo)))
+
+	ctx := context.Background()
+	c, err := cloudevents.NewClientHTTP(cloudevents.WithPort(env.Port))
+	if err != nil {
+		clog.Fatalf("failed to create event client, %v", err)
+	}
+	if err := c.StartReceiver(ctx, func(ctx context.Context, event cloudevents.Event) error {
+		// If this bot can handle the event, call the handler.
+
+		clog.FromContext(ctx).With("event", event).Debugf("received event")
+
+		defer func() {
+			if err := recover(); err != nil {
+				clog.Errorf("stacktrace from panic: %s", debug.Stack())
+			}
+		}()
+
+		if prb, ok := b.(interface {
```

Comment:
This is where I hid all the gross stuff. 🤢 

Because events we receive are `schema.PullRequestEvent`s and not full `*github.PullRequestEvent`s, we "cast" the former to the latter by JSON-round-tripping. This works, but it won't be pretty as we expand to more types. It also means bots can't act on a field that we don't explicitly add to `event_types.go` (as BigQuery types 😒), which also makes it recorded in BQ.